src/compile.rs

```diff
@@ -459,6 +459,197 @@ pub fn validate_rules(statements: &BTreeSet<ValidatedStatement>) -> Result<(), C
+                    }
+                    None => {
+                        is_term = false;
+                        break;
+                    }
+                }
+            }
+        }
+        if is_term {
+            term_ret_vec.push(func.get_cil_name().clone());
+        } else {
+            nonterm_ret_vec.push(func.clone());
+        }
+    }
+
+    (term_ret_vec, nonterm_ret_vec)
+}
+
+pub fn search_for_recursion(
+    terminated_list: &mut Vec<String>,
+    functions: &mut Vec<FunctionInfo>,
+) -> Result<(), CascadeErrors> {
+    let mut removed: u64 = 1;
+    while removed > 0 {
+        removed = 0;
+        for func in functions.clone().iter_mut() {
+            let mut is_term = true;
+            for call in func.original_body {
+                if let Statement::Call(call) = call {
+                    let mut call_cil_name = call.get_cil_name();
+                    // If we are calling something with this it must be in the same class
+                    // so hand place the class name
+                    if call_cil_name.contains("this-") {
+                        if let Some(class) = func.class {
+                            call_cil_name =
+                                call_cil_name.replace("this-", &(class.name.to_string() + "-"))
+                        } else {
+                            return Err(CascadeErrors::from(ErrorItem::make_compile_or_internal_error(
+                                "Could not determine class for 'this.' function call",
+                                Some(func.declaration_file),
+                                call.get_name_range(),
+                                "Perhaps you meant to place the function in a resource or domain?")));
+                        }
+                    }
+
+                    if terminated_list.contains(&call_cil_name) {
+                        continue;
+                    }
+                    is_term = false;
+                }
+            }
+            if is_term {
+                terminated_list.push(func.get_cil_name());
+                removed += 1;
+                let index = functions
+                    .iter()
+                    .position(|x| *x.get_cil_name() == func.get_cil_name())
+                    .unwrap();
+                functions.remove(index);
+            }
+        }
+    }
+
+    if !functions.is_empty() {
+        let mut error: Option<CompileError> = None;
+
+        for func in functions {
+            error = Some(add_or_create_compile_error(
+                error,
+                "Recursive Function call found",
+                func.declaration_file,
+                func.get_declaration_range().unwrap_or_default(),
+                "Calls cannot recursively call each other",
+            ));
+        }
+        // Unwrap is safe since we need to go through the loop above at least once
+        return Err(CascadeErrors::from(error.unwrap()));
+    }
+
+    Ok(())
+}
+
+pub fn prevalidate_functions(
```

Comment:
Would functions.rs be a better home for all of this?

Reply:
I put the prevalidate_functions method here since that's where validate_functions was.  Would you want me to move all of those functions(validate_functions, prevalidate_functions, the new functions prevalidate calls) or just the stuff that prevalidate_functions calls?

Reply:
Oh interesting...  I hadn't realized validate_functions() was here as well.  It does make sense for them to live together...

I think I'd really like both of them to ultimately migrate to functions.rs.  Maybe move the existing validate_functions.rs as its own commit as part of this PR.